src/java.base/share/classes/java/security/PEMEncoder.java

```diff
@@ -0,0 +1,358 @@
+ * on a PEMEncoder instance returned by {@link #withEncryption(char[])} or
+ * by passing an {@link EncryptedPrivateKeyInfo} object into the encode methods.
+ * <p>
+ * PKCS8 v2.0 allows OneAsymmetric encoding, which is a private and public
+ * key in the same PEM.  This is supported by using the {@link KeyPair} class
+ * with the encode methods.
+ * <p>
+ * PEMEncoder supports the follow types:
+ * <pre>
+ *     PRIVATE KEY, PUBLIC KEY, CERTIFICATE, CRL, and ENCRYPTED PRIVATE KEY.
+ * </pre>
+ *
+ * @apiNote
+ * Here is an example of encoding a PrivateKey object:
+ * <pre>
+ *     PEMEncoder pe = PEMEncoder.of();
+ *     byte[] pemData = pe.encode(privKey);
+ * </pre>
+ *
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+public final class PEMEncoder {
+
+    // Singleton instance of PEMEncoder
+    private static final PEMEncoder PEM_ENCODER = new PEMEncoder(null);
+
+    // Stores the password for an encrypted encoder that isn't setup yet.
+    private PBEKeySpec keySpec;
+    // Stores the key after the encoder is ready to encrypt.  The prevents
+    // repeated SecretKeyFactory calls if the encoder is used on multiple keys.
+    private SecretKey key;
+    // Makes SecretKeyFactory generation thread-safe.
+    private final ReentrantLock lock;
+    // Lazy initialize singleton encoder.
+    private static Base64.Encoder b64Encoder;
+
+    /**
+     * Instantiate a new PEMEncoder for Encrypted Private Keys.
+     *
+     * @param pbe contains the password spec used for encryption.
+     */
+    private PEMEncoder(PBEKeySpec pbe) {
+        keySpec = pbe;
+        lock = new ReentrantLock();
+    }
+
+    /**
+     * Returns an instance of PEMEncoder.
+     *
+     * @return PEMEncoder instance
+     */
+    static public PEMEncoder of() {
+        return PEM_ENCODER;
+    }
+
+    /**
+     * Construct a String-based encoding based off the id type.
+     * @return the string
+     */
+    private String pemEncoded(PEMRecord pem) {
+        StringBuilder sb = new StringBuilder(1024);
+        sb.append("-----BEGIN ").append(pem.id()).append("-----");
+        sb.append(System.lineSeparator());
+        if (b64Encoder == null) {
+            b64Encoder = Base64.getMimeEncoder(64,
+                System.lineSeparator().getBytes());
+        }
+        sb.append(b64Encoder.encodeToString(
+            pem.pem().getBytes(StandardCharsets.ISO_8859_1)));
+        sb.append(System.lineSeparator());
+        sb.append("-----END ").append(pem.id()).append("-----");
+        sb.append(System.lineSeparator());
+        return sb.toString();
+    }
+
+    /**
+     * Encoded a given {@code DEREncodable} and return the PEM encoding in a
+     * String
+     *
+     * @param de a cryptographic object to be PEM encoded that implements
+     *           DEREncodable.
+     * @return PEM encoding in a String
+     * @throws IllegalArgumentException when the passed object returns a null
```

Comment:
What does "returns a null binary encoding" mean? There is no other method talking about this. I think we can just say "if configured for encryption but object does not support" since this looks like the only reason.

Also, how about `IllegalStateException`?